zeebe/backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/FileSetManager.java

```diff
@@ -53,50 +54,62 @@ public FileSetManager(final S3AsyncClient client, final S3BackupConfig config) {
     // concurrent uploads to half of the number of available connections.
     // This should prevent ConnectionAcquisitionTimeout for backups with many and/or large files
     // where we would otherwise occupy all connections, preventing some uploads from starting.
-    uploadLimit = new Semaphore(Math.max(1, config.maxConcurrentConnections() / 2));
+    concurrencyLimit = new Semaphore(Math.max(1, config.maxConcurrentConnections() / 2));
+    threadBuilder = Thread.ofVirtual().name("zeebe-backup-", 0);
   }
 
   CompletableFuture<FileSet> save(final String prefix, final NamedFileSet files) {
     LOG.debug("Saving {} files to prefix {}", files.files().size(), prefix);
     return CompletableFutureUtils.mapAsync(
             files.namedFiles().entrySet(),
             Entry::getKey,
-            namedFile -> saveFile(prefix, namedFile.getKey(), namedFile.getValue()))
+            namedFile -> schedule(() -> saveFile(prefix, namedFile.getKey(), namedFile.getValue())))
         .thenApply(FileSet::new);
   }
 
-  private CompletableFuture<FileSet.FileMetadata> saveFile(
-      final String prefix, final String fileName, final Path filePath) {
+  /** Schedules a task to be executed asynchronously, respecting the concurrency limit. */
+  private <T> CompletableFuture<T> schedule(final Supplier<T> task) {
```

Comment:
:wrench: why not using a `Executors.newThreadPerTaskExecutor` and do a `Executor.submit`?

Reply:
I just realized that we don't keep a handle of the threads created, so we may have a leak.
I assume running virtual threads are not garbage collected as they are GC roots?